File: src/class_file.f90
module class_file

  use class_string, only : string

  implicit none
  
  private
  public :: file

  !-------------------------------------------------------------------!
  ! Derived type for file
  !-------------------------------------------------------------------!
  
  type :: file

     character(:), allocatable :: filename ! filename
     type(integer) :: file_unit            ! file unit number
     integer       :: buffer_size

   contains

     ! Override
     procedure :: open
     procedure :: close
     procedure :: get_unit
     procedure :: read_line
     procedure :: read_lines
     procedure :: get_num_lines

  end type file

  !-------------------------------------------------------------------!
  ! Interface to construct a file
  !-------------------------------------------------------------------!

  interface file
     module procedure create
  end interface file

contains

  !===================================================================!
  ! Construct a file object with filename
  !===================================================================!
  
  type(file) function create(filename, line_width) result (this)

    type(character(*)), intent(in)           :: filename
    type(integer)     , intent(in), optional :: line_width
    logical :: ok
    integer :: i

    ! Set the file name
    allocate(this % filename, source=filename)

    ! Line width
    if (present(line_width)) then
       this % buffer_size = line_width
    else
       this % buffer_size = 100
    end if

    ! Use an available handle for opening
    i = 99
    check_unit: do 
       i = i + 1 
       inquire(unit=i, opened=ok)
       if(ok .eqv. .false.) then
          this % file_unit = i
          exit check_unit
       end if
    end do check_unit

  end function create

  !===================================================================!
  ! Open the file
  !===================================================================!

  subroutine open(this)

    class(file), intent(in) :: this
    logical :: file_exists

    inquire(file=this % filename, exist=file_exists)
    if (file_exists  .eqv. .false.) then
       print *, 'file does not exist ', this % filename
       error stop
    end if
    
    open(unit = this % file_unit, file = this % filename)
    
  end subroutine open

  !===================================================================!
  ! Close the file
  !===================================================================!

  subroutine close(this)

    class(file), intent(in) :: this

    close(unit = this % file_unit)

  end subroutine close

  !===================================================================!
  ! return the file unit number
  !===================================================================!

  pure type(integer) function get_unit(this)

    class(file), intent(in) :: this

    get_unit = this % file_unit

  end function get_unit

  !===================================================================!
  ! Utility function for get number of lines in mesh file
  !===================================================================!

  type(integer) function get_num_lines(this) result(nlines)

    class(file) , intent(in) :: this
    integer :: stat

    nlines = 0 
    call this % open()
    do
       read(this % get_unit(),*, iostat=stat)
       if (stat .ne. 0) exit
       nlines = nlines + 1
    end do
    call this % close()

  end function get_num_lines

  !=================================================================!
  ! Read one line and return a string object
  !=================================================================!

  subroutine read_line(this, line)

    ! Arguments
    class(file)  , intent(in)  :: this
    type(string) , intent(out) :: line

    ! Locals
    character(len=this % buffer_size) :: buffer

    read(this % get_unit(), fmt = '(a)') buffer
    line = string(trim(buffer))

  end subroutine read_line

  !=================================================================!
  ! Read all lines and return a string array
  !=================================================================!

  subroutine read_lines(this, lines)

    ! Arguments
    class(file)  , intent(in)            :: this
    type(string) , allocatable, intent(out) :: lines(:)

    ! Locals
    integer :: num_lines
    integer :: iline

    ! Get number of lines in file to allocate space (uses a separate
    ! handle) fix!
    num_lines = this % get_num_lines()
    allocate(lines(num_lines))

    ! Loop through each line and read and store into lines
    call this % open()
    do iline = 1, num_lines
       call this % read_line(lines(iline))
    end do
    call this % close()

  end subroutine read_lines
   
end module class_file

File: ufvm/class_file.py
#=====================================================================#
# Contains a derived type 'file' and implemented procedures
#=====================================================================#

import os

from .class_string import String

# Module-level counter to emulate Fortran's free-unit discovery.
_next_unit = [100]


class File:
    """Derived type for file."""

    #===================================================================#
    # Construct a file object with filename
    #
    # Fortran: interface file => create(filename, line_width)
    #===================================================================#
    def __init__(self, filename, line_width=None):
        # Set the file name
        self.filename = filename

        # Line width
        if line_width is not None:
            self.buffer_size = line_width
        else:
            self.buffer_size = 100

        # Use an available handle for opening
        self.file_unit = _next_unit[0]
        _next_unit[0] += 1

        # Python file object backing the unit (None until opened)
        self._handle = None

    #===================================================================#
    # Open the file
    #===================================================================#
    def open(self):
        file_exists = os.path.exists(self.filename)
        if file_exists is False:
            print(' file does not exist ', self.filename)
            raise SystemExit
        self._handle = open(self.file_unit_to_path())

    def file_unit_to_path(self):
        # Helper: the unit maps to this file's path.
        return self.filename

    #===================================================================#
    # Close the file
    #===================================================================#
    def close(self):
        if self._handle is not None:
            self._handle.close()
            self._handle = None

    #===================================================================#
    # return the file unit number
    #===================================================================#
    def get_unit(self):
        return self.file_unit

    #===================================================================#
    # Utility function for get number of lines in mesh file
    #===================================================================#
    def get_num_lines(self):
        nlines = 0
        self.open()
        for _ in self._handle:
            nlines = nlines + 1
        self.close()
        return nlines

    #=================================================================#
    # Read one line and return a string object
    #=================================================================#
    def read_line(self):
        # Fortran reads a fixed-width buffer then trims trailing blanks.
        buffer = self._handle.readline()
        line = String(buffer.rstrip())
        return line

    #=================================================================#
    # Read all lines and return a string array
    #=================================================================#
    def read_lines(self):
        # Get number of lines in file to allocate space
        num_lines = self.get_num_lines()
        lines = [None] * num_lines

        # Loop through each line and read and store into lines
        self.open()
        for iline in range(num_lines):
            lines[iline] = self.read_line()
        self.close()
        return lines
